### PR TITLE
Add emscripten cache directory to default library path

### PR DESCRIPTION
**Pull request by sbc100**
Also, allow -L/-l flags to pass all the way down the linker under the
wasm backend.

This change paves that way to include system librayies such as -lc on
the link line by default rather than including them via the `.symbols`
files.

See #8912

### CODE REVIEW

File: tools/shared.py [hunk 1]
```diff
@@ -2737,7 +2758,8 @@ def path_to_system_js_libraries(library_name):
 
     elif library_name.endswith('.js') and os.path.isfile(path_from_root('src', 'library_' + library_name)):
       library_files += ['library_' + library_name]
-    else:
+    elif not Settings.WASM_BACKEND:
+      # The wasm backend will report these when wasm-ld runs
```
Comment (kripken) [inline tools/shared.py]: maybe we could have an assert on them here?

File: tools/shared.py [hunk 2]
```diff
@@ -965,38 +965,61 @@ def get_llvm_target():
...
+
+  if os.environ.get('EMMAKEN_NO_SDK') or '-nostdinc' in user_args:
+    return c_opts
+
   # Disable system C and C++ include directories, and add our own (using
-  # -idirafter so they are last, like system dirs, which allows projects to
+  # -isystem so they are last, like system dirs, which allows projects to
   # override them)
+
+  c_opts += ['-Xclang', '-nostdsysteminc']
```
Comment (kripken) [inline tools/shared.py]: empty line after this one, to match the ones later?

File: tests/test_other.py [hunk 3]
```diff
@@ -961,7 +961,7 @@ def test_wl_linkflags(self):
     -lfoo
     ''')
     run_process([PYTHON, EMCC, '-o', 'libfile.o', 'libfile.cpp'])
-    run_process([PYTHON, EMAR, 'rv', 'libfoo.a', 'libfile.o'])
+    run_process([PYTHON, EMAR, 'cr', 'libfoo.a', 'libfile.o'])
```
Comment (kripken) [inline tests/test_other.py]: why is this change needed?

File: emcc.py [hunk 4]
```diff
@@ -1910,6 +1913,8 @@ def compile_source_file(i, input_file):
         def supported(f):
           if f in SUPPORTED_LINKER_FLAGS:
             return True
+          if f.startswith('-l') or f.startswith('-L'):
+            return False
```
Comment (kripken) [inline emcc.py]: I don't understand this change?

File: emcc.py [hunk 5]
```diff
@@ -1801,7 +1803,8 @@ def check_human_readable_list(items):
 
       # Bitcode args generation code
       def get_clang_command(input_files):
-        args = [clang_compiler] + newargs + input_files
+        cflags = shared.COMPILER_OPTS + shared.emsdk_cflags(newargs)
```
Comment (kripken) [inline emcc.py]: for anything to work before, these would have had to be added somewhere, and then we can remove that place in this PR. I see that for `COMPILER_OPTS` but not for the other flags, so I'm missing something?

File: emcc.py [hunk 6]
```diff
@@ -1041,9 +1047,10 @@ def get_last_setting_change(setting):
     if not shared.Settings.STRICT:
       # The preprocessor define EMSCRIPTEN is deprecated. Don't pass it to code
       # in strict mode. Code should use the define __EMSCRIPTEN__ instead.
-      shared.COMPILER_OPTS += ['-DEMSCRIPTEN']
+      newargs += ['-DEMSCRIPTEN']
 
-    settings_changes.append(process_libraries(libs, lib_dirs, input_files))
+    consumed = process_libraries(libs, lib_dirs, input_files)
+    link_flags = [l for l in link_flags if l[0] not in consumed]
```
Comment (kripken) [inline emcc.py]: `l[0]` is the order number, maybe that should be `l[1]`?

Comment (sbc100) [reply emcc.py]: Yes, I'm trying to remove that flags that are not longer needed because they were consumed by process_libraries.  process_libraries returns a list of arg indexes.

File: emcc.py (hunk 5 above)
Comment (sbc100) [reply emcc.py]: Yes, in this PR COMPILER_OPTS is not longer used internally.  Its still there just because it can be set in the `.emscripten` file.. but really I think we should rid of it completely in a followup.

In any case COMPILER_OPTS used have all the sdk flags added to it in tools/shared.py.   That is no longer the case.

File: emcc.py (hunk 4 above)
Comment (sbc100) [reply emcc.py]: One of that things this change does is allow the user's -l and -L flags to propogate through to linker flags.  However, for fastcomp we want to strip these out here as they are handled elsewhere.

File: tests/test_other.py (hunk 3 above)
Comment (sbc100) [reply tests/test_other.py]: This is just cleanup.  Without this ar prints a warning that its creating an archive.

File: tools/shared.py (hunk 1 above)
Comment (sbc100) [reply tools/shared.py]: Can you explain?

The idea is that fastcomp errors out here on missing libraries.. but with the wasm-backend the error will happen later.

**Comment by sbc100** [comment]
Split out unrelated stuff into separate CL

File: tools/shared.py (hunk 1 above)
Comment (kripken) [reply tools/shared.py]: Oh, later, not earlier, I misunderstood. sgtm.
